wasm/validator/src/visitor.rs

```diff
@@ -0,0 +1,747 @@
+        (IdBox::TriggerId(id), Value::Numeric(NumericValue::U32(object))) => {
+            validator.validate_mint_trigger_repetitions(authority, &Mint::new(object, id))
+        }
+        _ => deny_unsupported_instruction!(Mint),
+    }
+}
+
+pub fn validate_burn<V: Validate + ?Sized>(
+    validator: &mut V,
+    authority: &AccountId,
+    isi: &BurnBox,
+) -> Verdict {
+    let destination_id = evaluate_field!(validator, authority, <isi as BurnBox>::destination_id);
+    let object = evaluate_field!(validator, authority, <isi as BurnBox>::object);
+
+    match (destination_id, object) {
+        (IdBox::AssetId(id), Value::Numeric(value)) => {
+            validator.validate_burn_asset(authority, &Burn::new(value, id))
+        }
+        (IdBox::AccountId(id), Value::PublicKey(value)) => {
+            validator.validate_burn_account_public_key(authority, &Burn::new(value, id))
+        }
+        _ => deny_unsupported_instruction!(Burn),
+    }
+}
+
+pub fn validate_transfer<V: Validate + ?Sized>(
+    validator: &mut V,
+    authority: &AccountId,
+    isi: &TransferBox,
+) -> Verdict {
+    let object = evaluate_field!(validator, authority, <isi as TransferBox>::object);
+
+    let (IdBox::AssetId(source_asset_id), IdBox::AssetId(destination_asset_id)) = (
+            evaluate_field!(validator, authority, <isi as TransferBox>::source_id),
+            evaluate_field!(validator, authority, <isi as TransferBox>::destination_id),
+        ) else {
+            deny_unsupported_instruction!(Transfer)
+        };
+
+    match object {
+        Value::Numeric(quantity) => validator.validate_transfer_asset(
+            authority,
+            &Transfer::new(source_asset_id, quantity, destination_asset_id),
+        ),
+        _ => deny_unsupported_instruction!(Transfer),
+    }
+}
+
+pub fn validate_set_key_value<V: Validate + ?Sized>(
+    validator: &mut V,
+    authority: &AccountId,
+    isi: &SetKeyValueBox,
+) -> Verdict {
+    let object_id = evaluate_field!(validator, authority, <isi as SetKeyValueBox>::object_id);
+    let key = evaluate_field!(validator, authority, <isi as SetKeyValueBox>::key);
+    let value = evaluate_field!(validator, authority, <isi as SetKeyValueBox>::value);
+
+    match object_id {
+        IdBox::AssetId(id) => {
+            validator.validate_set_asset_key_value(authority, &SetKeyValue::new(id, key, value))
+        }
+        IdBox::AssetDefinitionId(id) => validator
+            .validate_set_asset_definition_key_value(authority, &SetKeyValue::new(id, key, value)),
+        IdBox::AccountId(id) => {
+            validator.validate_set_account_key_value(authority, &SetKeyValue::new(id, key, value))
+        }
+        IdBox::DomainId(id) => {
+            validator.validate_set_domain_key_value(authority, &SetKeyValue::new(id, key, value))
+        }
+        _ => deny_unsupported_instruction!(SetKeyValue),
+    }
```

Comment:
I think we need a unified way of such matches. So that it won't be possible to update `core` and forget to update `default_validator`

Reply:
keep it in mind, this is something I'll address in the next PR

Reply:
Let's leave an issue or PR number here and then close this discussion

Reply:
done #3446